Add filter column names to query table list

stuartmcalpine opened the pull request:
Currently the "tables required" when doing a query are only built from the return column list. This means if a filter used a table not in the returned column list the proper join would not be made.

Added a new function and test to make sure the join is made in all cases.

JoanneBogart reviewed (approved):
This looks ok in itself, but it raised some other questions as I looked at related code .  These are comments I really should have made ages ago, when the code in question was added, so it's up to you if you want to address them in this PR or a separate one.

`_parse_selected_columns` doesn't describe returns, unfortunate since the returns really could use some explaining.  It took me a while to understand the purpose of
the last item, `is_orderable_list`.   And, now that I think I understand what it's for, I don't understand why the code at line 220 gives the right answer.  Maybe add a comment there?
Finally, in `_render_filter` there is an assert.   I'm not convinced we need it at all - how could it go wrong if f really is a filter as it's supposed to be? -  but if we do it should be changed to raise an exception.

stuartmcalpine commented:
> `_parse_selected_columns` doesn't describe returns, unfortunate since the returns really could use some explaining. It took me a while to understand the purpose of the last item, `is_orderable_list`. And, now that I think I understand what it's for, I don't understand why the code at line 220 gives the right answer. Maybe add a comment there? 

I've added some more info in the doc string about what's being returned

> Finally, in `_render_filter` there is an assert. I'm not convinced we need it at all - how could it go wrong if f really is a filter as it's supposed to be? - but if we do it should be changed to raise an exception.

I've moved the check to within `_parse_selected_columns` , just to make sure the lists come out the same size

JoanneBogart reviewed (approved):
This looks good.  Thanks.
